statsmodels/multivariate/factor.py

```diff
@@ -0,0 +1,330 @@
+# -*- coding: utf-8 -*-
+
+from statsmodels.base.model import Model
+from .factor_rotation import rotate_factors, promax
+
+try:
+    import matplotlib.pyplot as plt
+    missing_matplotlib = False
+except ImportError:
+    missing_matplotlib = True
+
+if not missing_matplotlib:
+    from .plots import plot_scree, plot_loadings
+
+import numpy as np
+from numpy.linalg import eigh, inv, norm, matrix_rank
+import pandas as pd
+
+from statsmodels.iolib import summary2
+
+
+class Factor(Model):
+    """
+    Factor analysis
+
+    .. [1] Hofacker, C. (2004). Exploratory Factor Analysis, Mathematical Marketing.
+    http://www.openaccesstexts.org/pdf/Quant_Chapter_11_efa.pdf
+
+    Supported rotations:
+        'varimax', 'quartimax', 'biquartimax', 'equamax', 'oblimin',
+        'parsimax', 'parsimony', 'biquartimin', 'promax'
+
+    Parameters
+    ----------
+    endog : array-like
+        Variables in columns, observations in rows
+        Could be `None` if `corr` is not `None`
+    n_factor : int
+        The number of factors to extract
+    corr : array-like
+        Directly specify the correlation matrix instead of estimating from endog
+        If not `None`, `endog` will not be used
+    method : str
+        Specify the method to extract factors
+        'pa' - Principal axis factor analysis
+    smc : True or False
+        Whether or not to apply squared multiple correlations
+    endog_names: str
+        Names of endogeous variables.
+        If specified, it will be used instead of the column names in endog
+
+    """
+    def __init__(self, endog, n_factor, corr=None, method='pa', smc=True,
+                 missing='drop', endog_names=None):
+        # CHeck validity of n_factor
+        if n_factor <= 0:
+            raise ValueError('n_factor must be larger than 0! %d < 0' %
+                             (n_factor))
+        if endog is not None and n_factor > endog.shape[1]:
+            raise ValueError('n_factor must be smaller or equal to the number'
+                             ' of columns of endog! %d > %d' %
+                             (n_factor, endog.shape[1]))
+        self.n_factor = n_factor
+
+        if corr is None and endog is None:
+            raise ValueError('Both endog and corr is None!')
+
+        self.loadings = None
+        self.communality = None
+        self.eigenvals = None
+        self.method = method
+        self.smc = smc
+
+        # Check validity of corr
+        if corr is not None:
+            if corr.shape[0] != corr.shape[1]:
+                raise ValueError('Correlation matrix corr must be a square')
+            if endog is not None and endog.shape[1] != corr.shape[0]:
+                    raise ValueError('The number of columns in endog must be '
+                                     'equal to the number of columns and rows corr')
+        self.corr = corr
+        self.endog_names = endog_names
```

Comment:
corr could be a pandas DataFrame
We need `self.corr = np.asarray(corr)` to make sure that we have an array for further processing.
endog_names could be taken from the index or column of the DataFrame,
e.g. `if hasattr(corr, index): endog_names=...`